src/app/qgisapp.cpp

```diff
@@ -9919,6 +9923,34 @@ void QgisApp::snappingOptions()
   mSnappingDialogContainer->show();
 }
 
+void QgisApp::enableDigitizeWithCurve( bool enable )
+{
+  mMapTools.mAddFeature->setCircularDigitizingEnable( enable );
+  static_cast<QgsMapToolCapture *>( mMapTools.mSplitFeatures )->setCircularDigitizingEnable( enable );
+  static_cast<QgsMapToolCapture *>( mMapTools.mReshapeFeatures )->setCircularDigitizingEnable( enable );
+  QgsSettings settings;
+  settings.setValue( QStringLiteral( "UI/digitizeWithCurve" ), enable ? 1 : 0 );
+}
+
+void QgisApp::enableDigitizeWithCurveAction( bool enable )
+{
+  QgsSettings settings;
+
+  QObject *sender = QObject::sender();
+  if ( sender && sender != this )
+    enable &= ( sender == mActionAddFeature && mMapTools.mAddFeature->mode() != QgsMapToolCapture::CapturePoint ) ||
+              sender == mActionSplitFeatures;
+
+  else
+    enable &= ( mMapCanvas->mapTool() == mMapTools.mAddFeature && mMapTools.mAddFeature->mode() != QgsMapToolCapture::CapturePoint ) ||
+              mMapCanvas->mapTool() == mMapTools.mSplitFeatures;
```

Comment:
why: 
because this feature is only available for adding new feature (line or polygon, not for point) and for splitting feature (could be possible in the future for reshape and add ring feature), so the action is enable (not the tool) and then, can be toggle, only if this is the current map tool is "add feature" (and not point) or "split feature".
And I find confusing for the user to have the action aways enable even if there is a tool that doesn't support curve digitizing, so much clearer for the user if the action is enabled only if a tool that supports digitizing with curve is activated.

How : 
action to toggle between curve/linear is enabled/disabled by the `QgisApp::enableDigitizeWithCurveAction` method.
This method `enableDigitizeWithCurveAction` can be called directly by the `QgisApp` instance (from `QgisApp::activateDeactivateLayerRelatedActions()` method, depending of active layer) or when actions `mActionAddFeature` or `mActionSplitFeatures` emit toggled signal.
When the method is called by `QgisApp`, sender is sometime null, sometime non null. In this case (`sender==null` or `sender == QgisApp`), `mActionDigitizeWithCurve` is enabled if the current map tool of the canvas is the add feature tool or the split feature tool and `enable==true`. Those case happen when QGis starts, when new vector layer are loaded or created or when active layer changes.
When the method is called through signal from `mActionAddFeature` or `mActionSplitFeatures`, `mActionDigitizeWithCurve` is enabled only if the sender is `mActionAddFeature` or `mActionAddFeature` and `enable==true`. This case happens when active tool changes.

If this behavior is not wanted, I can simply remove this code... There is no need to other mecanism because the digitizing with curve as to be enabled for each tools independently. So, if there is nothing to activate this for a specific map tool, it is simply not activated... Digitizing with curve is not like other as snapping or tracing, it is not general but independent for each tools that support this.

Reply:
We added `QgsMapToolCapture::Capabilities capabilities()` recently that must returns a "SupportsCurves" flag for tools that support curves, I think you can use that ?
(not really sure it applies as I haven't properly reviewed the whole PR)